src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.ProjectRestoreInfoSource.cs

```diff
@@ -172,52 +218,18 @@ protected virtual bool IsSavedNominationOutOfDate(ConfiguredProject activeConfig
             return false;
         }
 
-        protected virtual bool IsProjectConfigurationVersionOutOfDate(IReadOnlyCollection<PackageRestoreConfiguredInput> configuredInputs)
+        private void RegisterProjectRestoreInfoSource()
         {
-            Assumes.Present(_project.Services.ActiveConfiguredProjectProvider);
-            Assumes.Present(_project.Services.ActiveConfiguredProjectProvider.ActiveConfiguredProject);
-
-            if (configuredInputs is null)
-            {
-                return false;
-            }
-
-            var activeConfiguredProject = _project.Services.ActiveConfiguredProjectProvider.ActiveConfiguredProject;
-
-            IComparable? activeProjectConfigurationVersionFromConfiguredInputs = null;
-            foreach (var configuredInput in configuredInputs)
-            {
-                if (configuredInput.ProjectConfiguration.Equals(activeConfiguredProject.ProjectConfiguration))
-                {
-                    activeProjectConfigurationVersionFromConfiguredInputs = configuredInput.ConfiguredProjectVersion;
-                }
-            }
-
-            if (activeProjectConfigurationVersionFromConfiguredInputs is null ||
-                activeConfiguredProject.ProjectVersion.IsLaterThan(
-                    activeProjectConfigurationVersionFromConfiguredInputs))
-            {
-                return true;
-            }
-
-            if (configuredInputs.Count == 1)
-            {
-                return false;
-            }
-
-            foreach (var loadedProject in activeConfiguredProject.UnconfiguredProject.LoadedConfiguredProjects)
+            // Register before this project receives any data flows containing possible nominations.
+            // This is needed because we need to register before any nuget restore or before the solution load.
+#pragma warning disable RS0030 // Do not used banned APIs
+            var registerRestoreInfoSourceTask = Task.Run(async () =>
             {
-                foreach (var configuredInput in configuredInputs)
-                {
-                    if (loadedProject.ProjectConfiguration.Equals(configuredInput.ProjectConfiguration) &&
-                        loadedProject.ProjectVersion.IsLaterThan(configuredInput.ConfiguredProjectVersion))
-                    {
-                        return true;
-                    }
-                }
-            }
+                await _solutionRestoreService4.RegisterRestoreInfoSourceAsync(this, _projectAsynchronousTasksService.UnloadCancellationToken);
```

Comment:
Looks like this state machine can be avoided (i.e. remove `async` and `await`).

Reply:
Don.e